### PR TITLE
SAN-71 Pay via BACS selection page set up

### PR DESCRIPTION
**Pull request by klai-ch**
Set up Pay via BACS selection page

### CODE REVIEW

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        <p class="govuk-hint govuk-!-margin-top-3">For example, the penalty reference A1234567 starts with A.</p>
+        <div class="govuk-form-group" th:classappend="${#fields.hasErrors('selectedPenaltyReference')} ? 'govuk-form-group--error' : ''">
+          <fieldset class="govuk-fieldset">
+            <span class="govuk-error-message"
+                  id="selectedPaymentMethod-errorId"
+                  th:if="${#fields.hasErrors('selectedPenaltyReference')}"
+                  th:each="e : ${#fields.errors('selectedPenaltyReference')}" th:text="${e}">
+            </span>
+
+            <div class="govuk-radios" th:each="penaltyReference : ${availablePenaltyReference}" style="margin-bottom:1em;">
```
Comment (rbull-ch) [inline src/main/resources/templates/pps/bankTransferPenaltyReference.html:34]: I think we might need to add in `data-module="govuk-radios"` here

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+              <div class="govuk-radios__item">
+                <input class="govuk-radios__input piwik-event"
+                       th:id="${penaltyReference}"
+                       name="available-payment-method"
```
Comment (rbull-ch) [inline src/main/resources/templates/pps/bankTransferPenaltyReference.html:38]: Should this be `available-penalty-reference` instead?

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                       th:errorclass="govuk-error-message"
+                       th:attr="data-event-id=${penaltyReference}"/>
+                <label class="govuk-label govuk-radios__label" th:for="${penaltyReference}" th:text="${penaltyReference}">keyvalue</label>
+              </div>
+            </div>
+
+          </fieldset>
+        </div>
+        <div class="govuk-form-group">
+          <details class="govuk-details piwik-event" data-event-id="Details - Help with the penalty reference" data-module="govuk-details" aria-labelledby="penalty-reference">
```
Comment (rbull-ch) [inline src/main/resources/templates/pps/bankTransferPenaltyReference.html:51]: The `data-event-id="Details -` bit probably needs to be something like `data-event-id="Bank Transfer -`?

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+              <ul class="govuk-list govuk-list--bullet">
+                <li>your penalty reference does not start with any of these letters</li>
+                <li>you have lost your penalty notice - use 'Lost penalty notice' as the email subject, and tell us as much information as you can about the penalty so we can pass your message to the right team</li>
+              </ul>
+              <div id="penalty-notice-image"></div>
+            </div>
+          </details>
+        </div>
+        <div class="form-group">
+          <input id="next-button" class="govuk-button piwik-event" data-event-id="Penalty Reference - Continue" type="submit" role="button" value="Continue"/>
```
Comment (rbull-ch) [inline src/main/resources/templates/pps/bankTransferPenaltyReference.html:71]: The bit `data-event-id="Penalty Reference -` probably needs to be `data-event-id="Bank Transfer Penalty Reference -`?

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html [hunk 5]
```diff
@@ -0,0 +1,77 @@
+<!DOCTYPE html>
```
Comment (rbull-ch) [inline src/main/resources/templates/pps/bankTransferPenaltyReference.html:1]: Please can you change your HTML code style settings via `~/.editorconfig` preferably and reformat the html files you've added for consistency with other files in penalty-payment-web?

```
[*.html]
indent_size = 4
tab_width = 4
ij_continuation_indent_size = 8
```

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html (hunk 1 above)
Comment (klai-ch) [reply src/main/resources/templates/pps/bankTransferPenaltyReference.html]: Updated from the review suggested

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html (hunk 2 above)
Comment (klai-ch) [reply src/main/resources/templates/pps/bankTransferPenaltyReference.html]: Updated from the review suggested

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html (hunk 4 above)
Comment (klai-ch) [reply src/main/resources/templates/pps/bankTransferPenaltyReference.html]: Updated from the review suggested

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html (hunk 5 above)
Comment (klai-ch) [reply src/main/resources/templates/pps/bankTransferPenaltyReference.html:1]: Updated the indent part

File: src/main/java/uk/gov/companieshouse/web/pps/PPSWebApplication.java
Comment (rbull-ch) [inline src/main/java/uk/gov/companieshouse/web/pps/PPSWebApplication.java:1]: Please can you update line 15 to be `APPLICATION_NAME_SPACE = "penalty-payment-web";`?

**Review by rbull-ch — APPROVED** [review]
Approved with 1 comment - https://github.com/companieshouse/penalty-payment-web/pull/158/files#r1856457184

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    <title>
+        What type of penalty do you need to pay?
+    </title>
+</head>
+<div id="pps-penalty-reference" layout:fragment="content">
+    <form id="ppspenaltyreference" th:action="@{''}" th:object="${penaltyReferences}" method="post" class="form">
+        <div class="govuk-grid-row">
+            <div class="govuk-grid-column-two-thirds">
+
+                <div th:replace="fragments/ppsGlobalErrors :: globalErrors"></div>
```
Comment (rbull-ch) [inline src/main/resources/templates/pps/bankTransferPenaltyReference.html:17]: Please can you try updating this fragment expression with the `~{...}` envelope to remove the warning `[layouts/chsBaseLayout] Deprecated unwrapped fragment expression`?
See https://www.thymeleaf.org/doc/articles/thymeleaf31whatsnew.html

Please can you also update the `src/main/resources/templates/layouts/baseLayout.html` to have `serviceName = 'penalty-payment-web'`?

File: src/main/java/uk/gov/companieshouse/web/pps/models/AvailablePenaltyReference.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import uk.gov.companieshouse.web.pps.util.PenaltyReference;
+
+public class AvailablePenaltyReference {
+    private List<PenaltyReference> availablePenaltyReference;
+
+    public List<PenaltyReference> getAvailablePenaltyReference() {
+        availablePenaltyReference = new ArrayList<>();
+        availablePenaltyReference.add(PenaltyReference.LATE_FILING);
+        availablePenaltyReference.add(PenaltyReference.SANCTIONS);
+        return availablePenaltyReference;
```
Comment (acallaghan-ch) [inline src/main/java/uk/gov/companieshouse/web/pps/models/AvailablePenaltyReference.java:14]: Although I know this code is hit in one of the other tests, I would expect to see a unit test for this.

File: src/main/java/uk/gov/companieshouse/web/pps/models/PenaltyReferenceChoice.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.companieshouse.web.pps.models;
+
+import jakarta.validation.constraints.NotNull;
+
+public class PenaltyReferenceChoice {
+    @NotNull(message = "Select what the penalty reference starts with")
+    private String selectedPenaltyReference;
```
Comment (acallaghan-ch) [inline src/main/java/uk/gov/companieshouse/web/pps/models/PenaltyReferenceChoice.java:7]: Same here, I would expect a unit test. @dgordon-ch are you happy that other tests are hitting this code and that's enough?

File: src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import uk.gov.companieshouse.web.pps.util.PenaltyReference;
+
+@Controller
+@PreviousController(PPSStartController.class)
+@RequestMapping("/late-filing-penalty/bank-transfer/which-penalty-service")
+public class BankTransferPenaltyReferenceController extends BaseController {
+    private static final String PPS_BANK_TRANSFER_PENALTY_REFERENCE = "pps/bankTransferPenaltyReference";
+
+    private static final String PPS_BANK_TRANSFER_LATE_FILING_PATH = "/late-filing-penalty/bank-transfer/late-filing-details";
+    private static final String PPS_BANK_TRANSFER_SANCTIONS_PATH = "/late-filing-penalty/bank-transfer/sanctions-details";
```
Comment (acallaghan-ch) [inline src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java:28]: Sonar is saying to get these from a customisable parameter. Should we be getting these from a properties file @dgordon-ch?

Comment (dgordon-ch) [reply src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java:28]: In theory yes, anything we're using more than once especially. We might be able to leave this for a separate cleanup/debt task to remove that smell.

File: src/main/java/uk/gov/companieshouse/web/pps/models/PenaltyReferenceChoice.java (hunk 8 above)
Comment (dgordon-ch) [reply src/main/java/uk/gov/companieshouse/web/pps/models/PenaltyReferenceChoice.java:7]: I would say if the coverage exists upstream it should generally be fine for model/domain classes at least. Happy to cover it it's flagged by Sonar though.

File: src/main/java/uk/gov/companieshouse/web/pps/PPSWebApplication.java
Comment (dgordon-ch) [reply src/main/java/uk/gov/companieshouse/web/pps/PPSWebApplication.java:1]: Fyi, that was going to be included for changes on SAN-114, but we can do it here, no issue.

File: src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        if (bindingResult.hasErrors()) {
+            List<FieldError> errors = bindingResult.getFieldErrors();
+            for (FieldError error : errors) {
+                LOGGER.error(error.getObjectName() + " - " + error.getDefaultMessage());
+            }
+            model.addAttribute(PPS_AVAILABLE_PENALTY_REF_ATTR, getAvailablePenaltyReferenceDisplay());
+            return getTemplateName();
+        }
+
+        if (penaltyReferenceChoice.getSelectedPenaltyReference()
```
Comment (dgordon-ch) [inline src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java:63]:
```suggestion
        PenaltyReferenceChoice penaltyRefChoice = penaltyReferenceChoice.getSelectedPenaltyReference();

        if (penaltyRefChoice...
```

Comment (dgordon-ch) [reply src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java:63]: General practice, as we have the same get call in the else if too -- instantiate it first and use that.

File: src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                .equals(PenaltyReference.LATE_FILING.getPenaltyReference())){
+            return UrlBasedViewResolver.REDIRECT_URL_PREFIX + PPS_BANK_TRANSFER_LATE_FILING_PATH;
```
Comment (dgordon-ch) [inline src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java:65]: You could potentially do the same as above here with `UrlBasedViewResolver.REDIRECT_URL_PREFIX` -- instantiate and make it more readable, call it redirectUrlPrefix or something similar.

**Review by dgordon-ch — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java (hunk 10 above)
Comment (klai-ch) [reply src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java]: Updated, is it change like that?

File: src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java (hunk 11 above)
Comment (klai-ch) [reply src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java]: Updated, is it change like that?

File: src/main/resources/templates/pps/bankTransferPenaltyReference.html [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+                            <p> Email <a href="mailto:enquiries@companieshouse.gov.uk">enquiries@companieshouse.gov.uk</a> for help if:
+                            <ul class="govuk-list govuk-list--bullet">
+                                <li>your penalty reference does not start with any of these letters</li>
+                                <li>you have lost your penalty notice - use 'Lost penalty notice' as the email subject, and tell us as much information as you can about the penalty so we can pass your message to the right team</li></ul>
+                            <div id="penalty-notice-image"></div>
+                        </div>
+                    </details>
+                </div>
+                <div class="form-group">
+                    <input id="next-button" class="govuk-button piwik-event" data-event-id="Bank Transfer Penalty Reference - Continue" type="submit" role="button" value="Continue"/>
```
[the remark this answers is not in the document]
Comment (klai-ch) [reply src/main/resources/templates/pps/bankTransferPenaltyReference.html]: Update and remove from suggestion

File: src/main/java/uk/gov/companieshouse/web/pps/models/PenaltyReferenceChoice.java (hunk 8 above)
Comment (acallaghan-ch) [reply src/main/java/uk/gov/companieshouse/web/pps/models/PenaltyReferenceChoice.java:7]: Sonar isn't complaining, just me :) That's grand.

File: src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java (hunk 9 above)
Comment (acallaghan-ch) [reply src/main/java/uk/gov/companieshouse/web/pps/controller/pps/BankTransferPenaltyReferenceController.java:28]: Sounds good.

**Review by rbull-ch — APPROVED** [review]

**Pull request merged**